src/WalletConnectSigner.ts

```diff
@@ -0,0 +1,142 @@
+import * as ethers from "ethers";
+import * as ethersBytes from "@ethersproject/bytes";
+import type {
+  Provider,
+  TransactionRequest,
+  TransactionResponse,
+} from "@ethersproject/abstract-provider";
+
+import {
+  Deferrable,
+  defineReadOnly,
+  resolveProperties,
+} from "@ethersproject/properties";
+import { _TypedDataEncoder } from "ethers/lib/utils";
+import type { TypedDataDomain, TypedDataField } from "@ethersproject/abstract-signer";
+
+export class WalletConnectSigner extends ethers.Signer {
+  public walletConnect: WalletConnect;
+
+  private _provider: ethers.providers.JsonRpcProvider;
+  constructor(
+    walletConnect: WalletConnect,
+    provider: ethers.providers.JsonRpcProvider,
+  ) {
+    super();
+    defineReadOnly(this, "provider", provider);
+    this._provider = provider;
+    this.walletConnect = walletConnect;
+  }
+
+  async getAddress(): Promise<string> {
+    const accountAddress = this.walletConnect.accounts[0];
+    if (!accountAddress) {
+      throw new Error(
+        "The connected wallet has no accounts or there is a connection problem"
+      );
+    }
+    return accountAddress;
+  }
+
+
+  async _signTypedData(domain: TypedDataDomain, types: Record<string, Array<TypedDataField>>, value: Record<string, any>): Promise<string> {
+    // Populate any ENS names (in-place)
+    const populated = await _TypedDataEncoder.resolveNames(domain, types, value, (name: string) => {
+      return this._provider.resolveName(name);
+    });
+
+    const address = await this.getAddress();
+
+    const signature = await this.walletConnect.signTypedData([
+      address.toLowerCase(),
+      JSON.stringify(_TypedDataEncoder.getPayload(populated.domain, types, populated.value)),
+    ]);
+
+    return signature;
+  }
+
+  async signMessage(message: ethers.Bytes | string): Promise<string> {
+    const prefix = ethers.utils.toUtf8Bytes(
+      `\x19Ethereum Signed Message:\n${message.length}`
+    );
+    const msg = ethers.utils.concat([prefix, message]);
+    const address = await this.getAddress();
+    const keccakMessage = ethers.utils.keccak256(msg);
+    const signature = await this.walletConnect.signMessage([
+      address.toLowerCase(),
+      keccakMessage,
+    ]);
+    return signature;
+  }
+
+  async sendTransaction(
+    transaction: Deferrable<TransactionRequest>
+  ): Promise<TransactionResponse> {
+    const tx = await resolveProperties(transaction);
+    const from = tx.from || (await this.getAddress());
+
+    const txHash = await this.walletConnect.sendTransaction({
+      from,
+      to: tx.to,
+      value: maybeBigNumberToString(tx.value),
+      data: bytesLikeToString(tx.data),
+    });
+    return <TransactionResponse>{
+      hash: txHash,
+      nonce: tx.nonce,
+      gasLimit: tx.gasLimit,
+      gasPrice: tx.gasPrice,
+      data: bytesLikeToString(tx.data) || "",
+      value: tx.value,
+      chainId: tx.chainId,
+      confirmations: 1,
```

Comment:
Should be `0` no?

Reply:
yeah. that's the number confirmation that the provider uses to wait for the transaction before returning the TX reciept

Reply:
Should this be changed then? From the docs:

"The number of blocks that have been mined (including the initial block) since this transaction was mined."

Since no blocks are mined yet, it should be zero..